SB3 Serialization & Load Project

kchadha opened the pull request:
a.k.a Save/Load pt.2
### Resolves
Towards resolving #194.

### Proposed Changes
- SB3 Serialization cleanup (be explicit about what we're serializing, attempting to serialize only the minimum information necessary to load a project)
- Leverage changes in LLK/scratch-parser#21 to validate sb3 files
- Load .sb, .sb2, and .sb3 files (.sb files are still not supported, but they are at least passed through to scratch-parser)

#### Additional changes to come:
- More serialization work (serialize monitors, extension custom state)
- Support for sb1 import

#### Eventually (these are changes we want for the ideal world of save/load, but should not block release of this feature):
- Reducing unnecessary processes surrounding saving projects (only encode sounds/costumes when necessary instead of every time an edit is made -- even more eventually, move this to scratch-storage)
- Cleanup of unnecessary/duplicate information tracked in VM (e.g. assetId vs. md5 for sounds and costumes)
- Cleanup of relationship between scratch-vm, scratch-storage, and scratch-parser (this PR makes some progress towards this goal)

### Reason for Changes

Continuation Save/Load work.

### Test Coverage

Existing tests pass. Additional tests need to be added for sb3 round-trip testing.
(Travis build will fail until LLK/scratch-parser#21 gets merged)

### Additional Notes
This PR depends on LLK/scratch-parser#21, and is a dependency of LLK/scratch-gui#1621

kchadha committed:
Be specific about what we're serializing. Flatten targets first by turning them into JSON strings and re-parsing them, and then only serialize the properties necessary for saving and loading.

kchadha committed:
Always serialize parent and next on a block even if they're null.

kchadha committed:
Some more cleanup.

kchadha committed:
Major refactor of project loading. Removed loadProjectLocal, in favor of going back to single API (loadProject). loadProject uses scratch-parser to unpack and validate project. Scratch-parser is getting updated to return validated project, with appended project metadata, and unpacked zip (JSZip object) if provided a buffer.

kchadha committed:
Modifying apis to accomodate existing tests.

kchadha committed:
Remove todo

kchadha committed:
Update VM to use scratch-parser callback interface

kchadha committed:
Fix up tests.

cwillisf commented on src/serialization/sb3.js:
```diff
@@ -25,6 +25,87 @@ const {deserializeCostume, deserializeSound} = require('./deserialize-assets.js'
  */
 
+const serializeBlock = function (block) {
+    const obj = Object.create(null);
```
Interesting - I hadn't seen (noticed?) this before. Based on what I read [here](https://stackoverflow.com/questions/15518328/creating-js-object-with-object-createnull) maybe we should use this more often...

cwillisf commented on src/serialization/sb3.js:
```diff
@@ -25,6 +25,87 @@ const {deserializeCostume, deserializeSound} = require('./deserialize-assets.js'
+    obj.currentCostume = target.currentCostume;
+    obj.costumes = target.costumes.map(serializeCostume);
+    obj.sounds = target.sounds.map(serializeSound);
+    if (!obj.isStage) {
```
👍

cwillisf commented on src/virtual-machine.js:
```diff
@@ -177,34 +177,38 @@ class VirtualMachine extends EventEmitter {
+                    resolve(res);
+                }
+            });
+        });
```
If we end up doing this Node-style callback <=> Promise conversion more often it might make sense to make a utility function or two for it. I think this is the first one I've seen in VM, though, so I think this is fine for now.

cwillisf reviewed (approved):
LGTM, so long as you resolve the test failures of course

cwillisf commented:
(marking as needs work for test failure)

rschamp commented on src/serialization/sb3.js:
```diff
@@ -25,6 +25,87 @@ const {deserializeCostume, deserializeSound} = require('./deserialize-assets.js'
+const serializeBlocks = function (blocks) {
+    const obj = Object.create(null);
+    for (const blockID in blocks) {
+        obj[blockID] = serializeBlock(blocks[blockID]);
```
In `for...in` loops, we should use [hasOwnProperty](https://github.com/LLK/scratch-vm/blob/develop/src/blocks/scratch3_sound.js#L236) to make sure we're only looping on instance-level properties and not those added to the prototype (https://stackoverflow.com/questions/12735778/for-in-and-hasownproperty).

towerofnix replied:
Would it be better to just use `for...of` and Object.keys?

```js
for (const blockID of Object.keys(blocks)) {
  ...
}
```

rschamp commented on src/serialization/sb3.js:
```diff
@@ -25,6 +25,87 @@ const {deserializeCostume, deserializeSound} = require('./deserialize-assets.js'
+    obj.sampleCount = sound.sampleCount;
+    // TODO eventually want to get rid of this
+    obj.md5 = sound.md5;
+    // TODO do we need this soundID
```
I don't think so, since it's runtime-specific.

rschamp commented on src/serialization/sb3.js:
```diff
@@ -33,7 +114,12 @@ const {deserializeCostume, deserializeSound} = require('./deserialize-assets.js'
+    const flattenedOriginalTargets = JSON.parse(JSON.stringify(
+        runtime.targets.filter(target => target.isOriginal)));
+    obj.targets = flattenedOriginalTargets.map(t => serializeTarget(t, runtime));
+    // runtime.targets.filter(target => target.isOriginal);
```
Cruft?

rschamp commented on src/serialization/sb3.js:
```diff
@@ -114,7 +200,7 @@ const parseScratchObject = function (object, runtime, extensions, zip) {
             // fileUrl: soundSource.fileUrl,
```
Cruft?

rschamp commented on src/serialization/sb3.js:
```diff
@@ -114,7 +200,7 @@ const parseScratchObject = function (object, runtime, extensions, zip) {
             rate: soundSource.rate,
             sampleCount: soundSource.sampleCount,
-            soundID: soundSource.soundID,
+            // soundID: soundSource.soundID,
```
Cruft?

rschamp commented on src/virtual-machine.js:
```diff
@@ -177,34 +177,38 @@ class VirtualMachine extends EventEmitter {
-                                this.installTargets(targets, extensions, true));
-                    });
+        // Clear the current runtime
+        this.clear();
```
Maybe this should only happen once we're sure the input is going to work?

rschamp commented on src/virtual-machine.js:
```diff
@@ -258,47 +264,36 @@ class VirtualMachine extends EventEmitter {
     }
 
+    // TODO do we still need this function? Keeping it here so as not to introduce
+    // a breaking change.
```
Maybe at this point we should add a `log.warning()` when it's called?

rschamp commented on test/unit/serialization_sb3.js:
```diff
@@ -5,11 +5,13 @@ const demoSb3 = require('../fixtures/demo.json');
-    // @todo Analyze
-    t.type(JSON.stringify(result), 'string');
-    t.end();
+    vm.fromJSON(JSON.stringify(demoSb3))
```
This test should move to using `loadProject`.

rschamp reviewed (changes requested):
This looks great overall, just a few nits and questions.

kchadha committed:
Removing cruft...

kchadha committed:
When serializing md5+ext, the name should reflect that. Deserializing translates this name back to what vm and other components currently expect.

kchadha committed:
Remove cruft and fix for..in loops.

kchadha committed:
Clear runtime after validation success instead of before.

kchadha committed:
Add warning to 'fromJSON' function since it is now just a wrapper around loadProject. Change vm sb3_serialization test to use 'loadProject' function instead of 'fromJSON'.

kchadha commented:
@rschamp @cwillisf I believe I addressed all the PR comments asking for or suggesting changes. Please take another look and let me know what you think. These changes also involve the 'md5' to 'md5ext' name change during serialization (and then translating back the other way during deserialization). This change was discussed in the LLK/scratch-parser#21.

P.S. Tests will continue to fail until LLK/scratch-parser#21 is merged, and so this branch should remain un-merged until that happens and these tests finally pass.

rschamp reviewed (approved):
LGTM!

kchadha committed:
Check if loadProject input is any typed array using ArrayBuffer.isView. This allows us to pass a project's asset.data (from storage) directly instead of calling toString on it first.

cwillisf commented on src/virtual-machine.js:
```diff
@@ -182,14 +182,16 @@ class VirtualMachine extends EventEmitter {
      */
     loadProject (input) {
-        if (typeof input === 'object' && !(input instanceof ArrayBuffer)) {
+        if (typeof input === 'object' && !ArrayBuffer.isView(input)) {
```
Oh cool - I didn't know about `isView`!

kchadha replied:
@cwillisf Turns out ArrayBuffers are not views of themselves. Just added the instanceof check back in, but the isView check should work for Uint8Array, Float32Array, and all those good things.

kchadha committed:
Turns out ArrayBuffers are not views of themselves.

kchadha commented:
Tests pass now that scratch-parser has a new release on npm. Merging this PR.
